File: solutions/006-zigzag-conversion/solution.py
class Solution:
    def convert(self, s: str, num_rows: int) -> str:
        if num_rows == 1 or num_rows >= len(s):
            return s

        rows = [""] * num_rows
        cur_row = 0
        going_down = False

        for ch in s:
            rows[cur_row] += ch
            if cur_row == 0 or cur_row == num_rows - 1:
                going_down = not going_down
            cur_row += 1 if going_down else -1

        return "".join(rows)


if __name__ == "__main__":
    tests = [
        ("PAYPALISHIRING", 3, "PAHNAPLSIIGYIR"),
        ("PAYPALISHIRING", 4, "PINALSIGYAHRPI"),
        ("A", 1, "A"),
    ]
    sol = Solution()
    for s, num_rows, expected in tests:
        result = sol.convert(s, num_rows)
        status = "PASS" if result == expected else "FAIL"
        print(f"{status} | s={s!r}, numRows={num_rows} => {result!r} (expected {expected!r})")

File: solutions/006-zigzag-conversion/solution.go
package main

import "fmt"

// convert 模拟 Z 字形走法，按行收集字符
func convert(s string, numRows int) string {
	if numRows == 1 || numRows >= len(s) {
		return s
	}

	rows := make([][]byte, numRows)
	curRow := 0
	goingDown := false

	for i := 0; i < len(s); i++ {
		rows[curRow] = append(rows[curRow], s[i])
		if curRow == 0 || curRow == numRows-1 {
			goingDown = !goingDown
		}
		if goingDown {
			curRow++
		} else {
			curRow--
		}
	}

	out := make([]byte, 0, len(s))
	for _, row := range rows {
		out = append(out, row...)
	}
	return string(out)
}

func main() {
	tests := []struct {
		s, expected string
		numRows     int
	}{
		{"PAYPALISHIRING", "PAHNAPLSIIGYIR", 3},
		{"PAYPALISHIRING", "PINALSIGYAHRPI", 4},
		{"A", "A", 1},
	}

	for _, tc := range tests {
		result := convert(tc.s, tc.numRows)
		status := "PASS"
		if result != tc.expected {
			status = "FAIL"
		}
		fmt.Printf("%s | s=%q, numRows=%d => %q (expected %q)\n", status, tc.s, tc.numRows, result, tc.expected)
	}
}
